cmd/gardenlet/app/gardenlet.go

```diff
@@ -309,36 +327,26 @@ func NewGardenlet(ctx context.Context, cfg *config.GardenletConfiguration) (*Gar
 		return nil, fmt.Errorf("failed to get garden client: %w", err)
 	}
 
-	// Delete bootstrap auth data if certificate was freshly acquired
-	if len(csrName) > 0 {
+	// Delete bootstrap auth data if certificate was newly acquired
+	if len(csrName) > 0 && len(seedName) > 0 {
 		logger.Infof("Deleting bootstrap authentication data used to request a certificate")
 		if err := bootstrap.DeleteBootstrapAuth(ctx, k8sGardenClient.DirectClient(), csrName, seedName); err != nil {
 			return nil, err
 		}
 	}
 
-	seedRestCfg, err := kubernetes.RESTConfigFromClientConnectionConfiguration(&cfg.SeedClientConnection.ClientConnectionConfiguration, nil)
-	if err != nil {
-		return nil, err
-	}
-
 	// Set up leader election if enabled and prepare event recorder.
 	var (
 		leaderElectionConfig *leaderelection.LeaderElectionConfig
 		recorder             = utils.CreateRecorder(k8sGardenClient.Kubernetes(), "gardenlet")
 	)
 	if cfg.LeaderElection.LeaderElect {
-		k8sSeedClientLeaderElection, err := kubernetesclientset.NewForConfig(seedRestCfg)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create client for leader election: %w", err)
-		}
-
 		leaderElectionConfig, err = utils.MakeLeaderElectionConfig(
 			cfg.LeaderElection.LeaderElectionConfiguration,
 			*cfg.LeaderElection.LockObjectNamespace,
 			*cfg.LeaderElection.LockObjectName,
-			k8sSeedClientLeaderElection,
-			utils.CreateRecorder(k8sSeedClientLeaderElection, "gardenlet"),
+			seedClient.Kubernetes(),
```

Comment:
maybe one more time for me - why is it a problem if the tls bootstrapping, cert rotation and leader election have a common client?

Reply:
https://github.com/gardener/gardener/pull/2772#issuecomment-678984401

Reply:
> AFAIK that's a best practice in Kubernetes controllers to prevent the leader election contributing to the client-go's client-side rate limiting settings.

Without having tested it, it seems not relevant to me, as the TLS bootstrapping is only performed once during initial startup & the cert rotation once a year?

Reply:
I think the motivation is to prevent any unexpected side-effects when someone tunes the leader election flags (e.g., by setting a very high sync period that would consume all of the allowed QPS). I get your point (one-time actions), but if there is no reason to not create a dedicated clients I would still vote for doing it - just for the sake of a) following the best practices and b) preventing potential unnecessary bugs that we would need to chase/track down.

Reply:
there is now a dedicated client for the leader election + (TLS bootstrap & cert rotation)